STOR-1726: Minimal implementation of CIFS/Samba CSI driver operator

mpatlasov opened the pull request:
This PR implements the minimal set of go-files and yaml-assets to enable the command
```
./bin/csi-driver-smb-operator start --kubeconfig $KUBECONFIG --namespace openshift-cluster-csi-drivers
```
to install and run CIFS/Samba CSI driver:
```
$ oc get po -n openshift-cluster-csi-drivers
csi-smb-node-5hvqm                           3/3     Running   0          14s
csi-smb-node-f7mn2                           3/3     Running   0          14s
smb-csi-driver-controller-8657649897-dw47d   3/3     Running   0          16s
smb-csi-driver-controller-8657649897-fht45   3/3     Running   0          16s
```

@openshift/storage

mpatlasov commented:
/test e2e-azurestack-csi

mpatlasov commented:
/label docs-approved
/label px-approved

/assign @jsafrane

mpatlasov commented:
/test e2e-aws-ovn-upgrade

mpatlasov commented:
/test e2e-aws-ovn-upgrade

jsafrane commented on pkg/driver/samba/samba.go:
```diff
@@ -0,0 +1,45 @@
+
+// GetSambaFileOperatorControllerConfig returns second half of runtime configuration of the CSI driver operator,
+// after a client connection + cluster flavour are established.
+func GetSambaFileOperatorControllerConfig(ctx context.Context, flavour generator.ClusterFlavour, c *clients.Clients) (*config.OperatorControllerConfig, error) {
```
Just `GetSambaOperatorControllerConfig` ?

jsafrane commented on cmd/csi-driver-smb-operator/main.go:
Since most of objects have name `smb-csi-driver-*` and other operators are `xxx-csi-driver-operator`, I think the whole operator have `smb-csi-driver-operator` binary and be in `cmd/smb-csi-driver-operator`

jsafrane reviewed (commented):
It's a good start, I have couple of nits below.

mpatlasov commented:
Last forse-push integrated the following fixes:

 * `s/GetSambaFileOperatorControllerConfig/GetSambaOperatorControllerConfig/g`
 * rename cmd subdir from `csi-driver-smb-operator` to `smb-csi-driver-operator`
 * added `privileged: true` to smb container in `controller.yaml`
 * added ClusterRoleBinding to allow the controller to run privileged pods (`controller_privileged_binding.yaml`)
 * added ClusterRoleBinding to allow the provisioner to read any secret in the cluster (`configmap_and_secret_reader_provisioner_binding.yaml`)

This made dynamic provisioning work for me.

jsafrane commented on cmd/smb-csi-driver-operator/main.go:
```diff
@@ -0,0 +1,49 @@
+
+func NewOperatorCommand() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "csi-driver-smb-operator",
```
smb-csi-driver-operator

jsafrane commented on cmd/smb-csi-driver-operator/main.go:
```diff
@@ -0,0 +1,49 @@
+	}
+
+	ctrlCmd := controllercmd.NewControllerCommandConfig(
+		"csi-driver-smb-operator",
```
smb-csi-driver-operator

jsafrane commented on cmd/smb-csi-driver-operator/main.go:
```diff
@@ -0,0 +1,49 @@
+}
+
+func runCSIDriverOperator(ctx context.Context, controllerConfig *controllercmd.ControllerContext) error {
+	opConfig := smb.GetSambaFileOperatorConfig()
```
GetSambaOperatorConfig

jsafrane commented on pkg/driver/samba/samba.go:
```diff
@@ -0,0 +1,45 @@
+func GetSambaFileOperatorConfig() *config.OperatorConfig {
+	return &config.OperatorConfig{
+		CSIDriverName:                   opv1.SambaCSIDriver,
+		UserAgent:                       "csi-driver-smb-operator",
```
smb-csi-driver-operator

mpatlasov commented:
Last force-push made couple of renames in `pkg/driver/samba/samba.go` and `cmd/smb-csi-driver-operator/main.go`:

 * `s/csi-driver-smb-operator/smb-csi-driver-operator`
 * `s/GetSambaFileOperatorConfig/GetSambaOperatorConfig`

jsafrane commented:
/lgtm
/label qe-approved
We will do post merge testing, when the operator is compete.

mpatlasov commented:
/retest-required

Phaow commented:
/retest-required
